// @forestadmin/datasource-sql/dist/orm-builder/helpers/sequelize-type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable max-len */
/* eslint-disable max-classes-per-file */
const sequelize_1 = require("sequelize");
class SequelizeTypeFactory {
    static makeType(dialect, type, table, columnName) {
        switch (type.type) {
            case 'scalar':
                if (sequelize_1.DataTypes[type.subType])
                    return sequelize_1.DataTypes[type.subType];
                throw new Error(`Unexpected type: ${type.subType}`);
            case 'enum':
                // Use a custom type only if the name is not the default one.
                // This should prevent side-effects on most cases if the custom type fails to mimic the
                // default one and cause issues, while still allowing to use custom types when required.
                return dialect === 'postgres' && type.name && type.name !== `enum_${table}_${columnName}`
                    ? this.makeCustomEnumType(type.schema, type.name, type.values)
                    : sequelize_1.DataTypes.ENUM(...type.values);
            case 'array':
                return sequelize_1.DataTypes.ARRAY(this.makeType(dialect, type.subType, table, columnName));
            default:
                throw new Error('Invalid type');
        }
    }
    /**
     * This workaround is needed because when inserting records in a table with an Array<Enum> column,
     * sequelize will try to cast the value to the enum type.
     *
     * This will fail if the database that sequelize is connected to was not initially created by
     * sequelize, because the enum type is unlikely to have the name that sequelize expects.
     *
     * @see https://community.forestadmin.com/t/postgresql-enum-type-does-not-exist-error-wrong-enum-type-name/5931
     * @see https://github.com/sequelize/sequelize/blob/v6.28.0/src/dialects/postgres/data-types.js#L491
     * @see https://github.com/sequelize/sequelize/blob/v6.28.0/src/utils.js#L555
     */
    static makeCustomEnumType(schema, name, values) {
        var _a;
        const key = `${schema ? `"${schema}".` : ''}"${name}"`;
        const Type = (_a = class extends sequelize_1.DataTypes.ABSTRACT {
                constructor() {
                    super();
                    this.isDataSourceSqlEnum = true;
                    this.key = key;
                    // Steal the validate method from the ENUM type
                    this.validate = sequelize_1.DataTypes.ENUM.prototype.validate;
                    this.values = values;
                    this.options = { values };
                }
            },
            // Markers to tell @forestadmin/datasource-sequelize to consider this type as an enum
            // when transforming sequelize models to forest collections.
            // Otherwise, we would get the "Skipping column" error
            _a.isDataSourceSqlEnum = true,
            // Setting this tells sequelize the name of the type in the database.
            // This is used, most notably, when casting values (which happens when the enum is used in
            // arrays)
            _a.key = key,
            _a);
        return new Type();
    }
}
exports.default = SequelizeTypeFactory;
