// @forestadmin/datasource-sequelize/dist/utils/aggregation.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const datasource_toolkit_1 = require("@forestadmin/datasource-toolkit");
const date_aggregation_converter_1 = __importDefault(require("./date-aggregation-converter"));
const serializer_1 = __importDefault(require("./serializer"));
const un_ambigous_1 = __importDefault(require("./un-ambigous"));
class AggregationUtils {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    constructor(model) {
        this.aggregateFieldName = '__aggregate__';
        this.model = model;
        this.dialect = this.model.sequelize.getDialect();
        this.col = this.model.sequelize.col;
        this.dateAggregationConverter = new date_aggregation_converter_1.default(this.model.sequelize);
    }
    getGroupFieldName(groupField) {
        return `${groupField}__grouped__`;
    }
    quoteField(field) {
        try {
            const safeField = (0, un_ambigous_1.default)(this.model, field, true);
            return this.model.sequelize.getQueryInterface().quoteIdentifiers(safeField);
        }
        catch {
            throw new datasource_toolkit_1.ValidationError(`Invalid access: "${field}" on "${this.model.name}" does not exist.`);
        }
    }
    getGroupAndAttributesFromAggregation(aggregationQueryGroup) {
        const attributes = [];
        const groups = aggregationQueryGroup?.map(group => {
            const { field } = group;
            const groupFieldName = this.getGroupFieldName(field);
            const groupField = this.quoteField(field);
            if (group.operation) {
                const groupFunction = this.dateAggregationConverter.convertToDialect(groupField, group.operation);
                attributes.push([groupFunction, groupFieldName]);
                return this.dialect === 'mssql' ? groupFunction : groupFieldName;
            }
            attributes.push([this.col(groupField), groupFieldName]);
            return this.dialect === 'mssql' ? groupField : groupFieldName;
        });
        return { groups, attributes };
    }
    getOrder(aggregationFunction) {
        let order;
        // FIXME handle properly order
        switch (this.dialect) {
            case 'postgres':
                order = [this.col(this.aggregateFieldName), 'DESC NULLS LAST'];
                break;
            case 'mssql':
                order = [aggregationFunction, 'DESC'];
                break;
            default:
                order = [this.col(this.aggregateFieldName), 'DESC'];
        }
        return order;
    }
    computeResult(rows, groups, expectsNumber) {
        return rows.map(row => {
            let value = row[this.aggregateFieldName];
            // Workaround Sequelize casting sums to strings.
            // This happens since sequelize@6.27.0 because sequelize implemented support for bigints.
            if (expectsNumber && typeof value === 'string' && !Number.isNaN(Number(value)))
                value = Number(value);
            return {
                value,
                group: (groups ?? []).reduce((memo, { field }) => {
                    memo[field] = serializer_1.default.serializeValue(row[this.getGroupFieldName(field)]);
                    return memo;
                }, {}),
            };
        });
    }
}
exports.default = AggregationUtils;
