// @forestadmin/datasource-sequelize/dist/utils/query-converter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const datasource_toolkit_1 = require("@forestadmin/datasource-toolkit");
const sequelize_1 = require("sequelize");
const un_ambigous_1 = __importDefault(require("./un-ambigous"));
class QueryConverter {
    constructor(model) {
        this.model = model;
        this.dialect = this.model.sequelize.getDialect();
        this.col = this.model.sequelize.col;
        this.fn = this.model.sequelize.fn;
        this.where = this.model.sequelize.where;
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    makeWhereClause(field, operator, value) {
        switch (operator) {
            // Presence
            case 'Present':
                return { [sequelize_1.Op.ne]: null };
            case 'Missing':
                return { [sequelize_1.Op.is]: null };
            // Equality
            case 'Equal':
                return { [value !== null ? sequelize_1.Op.eq : sequelize_1.Op.is]: value };
            case 'NotEqual':
                return { [sequelize_1.Op.ne]: value };
            case 'In':
                return this.makeInWhereClause(field, value);
            case 'NotIn':
                return this.makeNotInWhereClause(field, value);
            // Orderables
            case 'LessThan':
                return { [sequelize_1.Op.lt]: value };
            case 'GreaterThan':
                return { [sequelize_1.Op.gt]: value };
            // Strings
            case 'Like':
                return this.makeLikeWhereClause(field, value, true, false);
            case 'ILike':
                return this.makeLikeWhereClause(field, value, false, false);
            case 'NotContains':
                return this.makeLikeWhereClause(field, `%${value}%`, true, true);
            // Arrays
            case 'IncludesAll':
                return { [sequelize_1.Op.contains]: Array.isArray(value) ? value : [value] };
            default:
                throw new Error(`Unsupported operator: "${operator}".`);
        }
    }
    makeInWhereClause(field, value) {
        const valueAsArray = value;
        if (valueAsArray.length === 1) {
            return this.makeWhereClause(field, 'Equal', valueAsArray[0]);
        }
        if (valueAsArray.includes(null)) {
            const valueAsArrayWithoutNull = valueAsArray.filter(v => v !== null);
            return {
                [sequelize_1.Op.or]: [this.makeInWhereClause(field, valueAsArrayWithoutNull), { [sequelize_1.Op.is]: null }],
            };
        }
        return { [sequelize_1.Op.in]: valueAsArray };
    }
    makeNotInWhereClause(field, value) {
        const valueAsArray = value;
        if (valueAsArray.length === 1) {
            return this.makeWhereClause(field, 'NotEqual', valueAsArray[0]);
        }
        if (valueAsArray.includes(null)) {
            const valueAsArrayWithoutNull = valueAsArray.filter(v => v !== null);
            return {
                [sequelize_1.Op.and]: [{ [sequelize_1.Op.ne]: null }, this.makeNotInWhereClause(field, valueAsArrayWithoutNull)],
            };
        }
        return { [sequelize_1.Op.notIn]: valueAsArray };
    }
    makeLikeWhereClause(field, value, caseSensitive, not) {
        const op = not ? 'NOT LIKE' : 'LIKE';
        const seqOp = not ? sequelize_1.Op.notLike : sequelize_1.Op.like;
        if (caseSensitive) {
            if (this.dialect === 'sqlite') {
                const sqLiteOp = not ? 'NOT GLOB' : 'GLOB';
                return this.where(this.col(field), sqLiteOp, value.replace(/%/g, '*').replace(/_/g, '?'));
            }
            if (this.dialect === 'mysql' || this.dialect === 'mariadb')
                return this.where(this.fn('BINARY', this.col(field)), op, value);
            return { [seqOp]: value };
        }
        if (this.dialect === 'postgres')
            return { [sequelize_1.Op.iLike]: value };
        if (this.dialect === 'mysql' || this.dialect === 'mariadb' || this.dialect === 'sqlite')
            return { [seqOp]: value };
        return this.where(this.fn('LOWER', this.col(field)), op, value.toLocaleLowerCase());
    }
    /*
     * Delete and update sequelize methods does not provide the include options.
     * This method is developed to by pass this problem.
     */
    async getWhereFromConditionTreeToByPassInclude(conditionTree) {
        const include = conditionTree
            ? this.getIncludeFromProjection(new datasource_toolkit_1.Projection(), conditionTree.projection)
            : [];
        const whereOptions = this.getWhereFromConditionTree(conditionTree);
        if (include.length === 0) {
            return whereOptions;
        }
        const keys = [...this.model.primaryKeyAttributes];
        const records = await this.model.findAll({ attributes: keys, where: whereOptions, include });
        const conditions = records.map(record => {
            const equals = keys.map(pk => new datasource_toolkit_1.ConditionTreeLeaf(pk, 'Equal', record.get(pk)));
            return datasource_toolkit_1.ConditionTreeFactory.intersect(...equals);
        });
        const union = datasource_toolkit_1.ConditionTreeFactory.union(...conditions);
        return this.getWhereFromConditionTree(union);
    }
    getWhereFromConditionTree(conditionTree) {
        if (!conditionTree)
            return {};
        const sequelizeWhereClause = {};
        if (conditionTree.aggregator !== undefined) {
            const { aggregator, conditions } = conditionTree;
            if (aggregator === null) {
                throw new Error('Invalid (null) aggregator.');
            }
            const sequelizeOperator = aggregator === 'And' ? sequelize_1.Op.and : sequelize_1.Op.or;
            if (!Array.isArray(conditions)) {
                throw new Error('Conditions must be an array.');
            }
            sequelizeWhereClause[sequelizeOperator] = conditions.map(condition => this.getWhereFromConditionTree(condition));
        }
        else if (conditionTree.operator !== undefined) {
            const { field, operator, value } = conditionTree;
            const isRelation = field.includes(':');
            const safeField = (0, un_ambigous_1.default)(this.model, field);
            sequelizeWhereClause[isRelation ? `$${safeField}$` : safeField] = this.makeWhereClause(safeField, operator, value);
        }
        else {
            throw new Error('Invalid ConditionTree.');
        }
        return sequelizeWhereClause;
    }
    getIncludeFromProjection(attrProjection, tableProjection = new datasource_toolkit_1.Projection()) {
        const projection = attrProjection.union(tableProjection);
        return Object.keys(projection.relations).map(name => {
            const relAttrProjection = attrProjection.relations[name] ?? new datasource_toolkit_1.Projection();
            const relTableProjection = tableProjection.relations[name] ?? new datasource_toolkit_1.Projection();
            return {
                association: name,
                attributes: relAttrProjection.columns,
                include: this.getIncludeFromProjection(relAttrProjection, relTableProjection),
            };
        });
    }
    getOrderFromSort(sort) {
        return (sort ?? []).map(({ field, ascending }) => {
            const path = field.split(':');
            return [...path, ascending ? 'ASC' : 'DESC'];
        });
    }
}
exports.default = QueryConverter;
