// @forestadmin/datasource-customizer/dist/decorators/actions/collection.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const datasource_toolkit_1 = require("@forestadmin/datasource-toolkit");
const base_1 = __importDefault(require("./context/base"));
const single_1 = __importDefault(require("./context/single"));
const result_builder_1 = __importDefault(require("./result-builder"));
class ActionCollectionDecorator extends datasource_toolkit_1.CollectionDecorator {
    constructor() {
        super(...arguments);
        this.actions = {};
    }
    addAction(name, action) {
        this.actions[name] = action;
        this.markSchemaAsDirty();
    }
    async execute(caller, name, data, filter) {
        const action = this.actions[name];
        if (!action)
            return this.childCollection.execute(caller, name, data, filter);
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const context = this.getContext(caller, action, data, filter);
        const resultBuilder = new result_builder_1.default();
        const result = await action.execute(context, resultBuilder);
        return (result || {
            type: 'Success',
            invalidated: new Set(),
            message: 'Success',
        });
    }
    async getForm(caller, name, data, filter, metas) {
        const action = this.actions[name];
        if (!action)
            return this.childCollection.getForm(caller, name, data, filter, metas);
        if (!action.form)
            return [];
        const formValues = data ? { ...data } : {};
        const used = new Set();
        const context = this.getContext(caller, action, formValues, filter, used, metas?.changedField);
        // Convert DynamicField to ActionField in successive steps.
        let dynamicFields = action.form.map(c => ({ ...c }));
        if (metas?.searchField) {
            // in the case of a search hook,
            // we don't want to rebuild all the fields. only the one searched
            dynamicFields = [dynamicFields.find(field => field.label === metas.searchField)];
        }
        dynamicFields = await this.dropDefaults(context, dynamicFields, formValues);
        dynamicFields = await this.dropIfs(context, dynamicFields);
        const fields = await this.dropDeferred(context, metas?.searchValues, dynamicFields);
        for (const field of fields) {
            // customer did not define a handler to rewrite the previous value => reuse current one.
            if (field.value === undefined)
                field.value = formValues[field.label];
            // fields that were accessed through the context.formValues.X getter should be watched.
            field.watchChanges = used.has(field.label);
        }
        return fields;
    }
    refineSchema(subSchema) {
        const newSchema = { ...subSchema, actions: { ...subSchema.actions } };
        for (const [name, { form, scope, generateFile }] of Object.entries(this.actions)) {
            // An action form can be send in the schema to avoid calling the load handler
            // as long as there is nothing dynamic in it.
            const isDynamic = form?.some(field => Object.values(field).some(value => typeof value === 'function') ||
                // A field with a hardcoded file should not be sent to the apimap. it is marked dynamic
                (field.type.includes('File') && field.defaultValue));
            newSchema.actions[name] = { scope, generateFile: !!generateFile, staticForm: !isDynamic };
        }
        return newSchema;
    }
    getContext(caller, action, formValues, filter, used, changedField) {
        return new {
            Global: base_1.default,
            Bulk: base_1.default,
            Single: single_1.default,
        }[action.scope](this, caller, formValues, filter, used, changedField);
    }
    async dropDefaults(context, fields, data) {
        const unvaluedFields = fields.filter(field => data[field.label] === undefined);
        const defaults = await Promise.all(unvaluedFields.map(field => this.evaluate(context, null, field.defaultValue)));
        unvaluedFields.forEach((field, index) => {
            data[field.label] = defaults[index];
        });
        fields.forEach(field => delete field.defaultValue);
        return fields;
    }
    async dropIfs(context, fields) {
        // Remove fields which have falsy if
        const ifValues = await Promise.all(fields.map(field => !field.if || this.evaluate(context, null, field.if)));
        const newFields = fields.filter((_, index) => ifValues[index]);
        newFields.forEach(field => delete field.if);
        return newFields;
    }
    async dropDeferred(context, searchValues, fields) {
        const newFields = fields.map(async (field) => {
            const keys = Object.keys(field);
            const values = await Promise.all(Object.values(field).map(value => this.evaluate(context, searchValues?.[field.label], value)));
            return keys.reduce((memo, key, index) => ({ ...memo, [key]: values[index] }), {});
        });
        return Promise.all(newFields);
    }
    async evaluate(context, searchValue, value) {
        if (this.isHandler(value)) {
            // Only the options key of the dynamic search dropdown widget accept a searchValue
            if (this.isSearchOptionsHandler(value)) {
                return value(context, searchValue);
            }
            return value(context);
        }
        return value;
    }
    // eslint-disable-next-line @typescript-eslint/ban-types
    isHandler(value) {
        return typeof value === 'function';
    }
    isSearchOptionsHandler(value) {
        return value.name === 'options';
    }
}
exports.default = ActionCollectionDecorator;
