// @forestadmin/datasource-toolkit/dist/src/decorators/collection-decorator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class CollectionDecorator {
    get nativeDriver() {
        return this.childCollection.nativeDriver;
    }
    get schema() {
        if (!this.lastSchema) {
            // If the schema is not cached (at the first call, or after a markSchemaAsDirty call),
            const subSchema = this.childCollection.schema;
            this.lastSchema = this.refineSchema(subSchema);
        }
        return this.lastSchema;
    }
    get name() {
        return this.childCollection.name;
    }
    constructor(childCollection, dataSource) {
        this.childCollection = childCollection;
        this.dataSource = dataSource;
        // When the child collection invalidates its schema, we also invalidate ours.
        // This is done like this, and not in the markSchemaAsDirty method, because we don't have
        // a reference to parent collections from children.
        if (childCollection instanceof CollectionDecorator) {
            const originalChildMarkSchemaAsDirty = childCollection.markSchemaAsDirty;
            childCollection.markSchemaAsDirty = () => {
                // Call the original method (the child)
                originalChildMarkSchemaAsDirty.call(childCollection);
                // Invalidate our schema (the parent)
                this.markSchemaAsDirty();
            };
        }
    }
    async execute(caller, name, data, filter) {
        const refinedFilter = await this.refineFilter(caller, filter);
        return this.childCollection.execute(caller, name, data, refinedFilter);
    }
    async getForm(caller, name, data, filter, metas) {
        const refinedFilter = await this.refineFilter(caller, filter);
        return this.childCollection.getForm(caller, name, data, refinedFilter, metas);
    }
    async create(caller, data) {
        return this.childCollection.create(caller, data);
    }
    async list(caller, filter, projection) {
        const refinedFilter = await this.refineFilter(caller, filter);
        return this.childCollection.list(caller, refinedFilter, projection);
    }
    async update(caller, filter, patch) {
        const refinedFilter = await this.refineFilter(caller, filter);
        return this.childCollection.update(caller, refinedFilter, patch);
    }
    async delete(caller, filter) {
        const refinedFilter = await this.refineFilter(caller, filter);
        return this.childCollection.delete(caller, refinedFilter);
    }
    async aggregate(caller, filter, aggregation, limit) {
        const refinedFilter = await this.refineFilter(caller, filter);
        return this.childCollection.aggregate(caller, refinedFilter, aggregation, limit);
    }
    async renderChart(caller, name, recordId) {
        return this.childCollection.renderChart(caller, name, recordId);
    }
    markSchemaAsDirty() {
        this.lastSchema = null;
    }
    async refineFilter(caller, filter) {
        return filter;
    }
    refineSchema(subSchema) {
        return subSchema;
    }
}
exports.default = CollectionDecorator;
