// @forestadmin/datasource-sequelize/dist/collection.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const datasource_toolkit_1 = require("@forestadmin/datasource-toolkit");
const sequelize_1 = require("sequelize");
const aggregation_1 = __importDefault(require("./utils/aggregation"));
const error_handler_1 = __importDefault(require("./utils/error-handler"));
const model_to_collection_schema_converter_1 = __importDefault(require("./utils/model-to-collection-schema-converter"));
const query_converter_1 = __importDefault(require("./utils/query-converter"));
const serializer_1 = __importDefault(require("./utils/serializer"));
class SequelizeCollection extends datasource_toolkit_1.BaseCollection {
    constructor(name, datasource, 
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    model, logger) {
        if (!model)
            throw new Error('Invalid (null) model instance.');
        super(name, datasource, {
            sequelize: model.sequelize,
            model,
            rawQuery: async (sql, replacements) => {
                const result = await model.sequelize.query(sql, {
                    type: sequelize_1.QueryTypes.RAW,
                    plain: false,
                    raw: true,
                    replacements,
                });
                return result?.[0];
            },
        });
        this.model = model;
        this.col = this.model.sequelize.col;
        this.fn = this.model.sequelize.fn;
        this.aggregationUtils = new aggregation_1.default(this.model);
        this.queryConverter = new query_converter_1.default(this.model);
        const modelSchema = model_to_collection_schema_converter_1.default.convert(this.model, logger);
        this.enableCount();
        this.addFields(modelSchema.fields);
        this.addSegments(modelSchema.segments);
    }
    async create(caller, data) {
        const records = await (0, error_handler_1.default)('create', () => this.model.bulkCreate(data));
        return records.map(record => serializer_1.default.serialize(record.get({ plain: true })));
    }
    async list(caller, filter, projection) {
        const query = {
            attributes: projection.columns,
            where: this.queryConverter.getWhereFromConditionTree(filter.conditionTree),
            include: this.queryConverter.getIncludeFromProjection(
            // Bugfix [CU-860rc94dq](https://app.clickup.com/t/860rc94dq)
            // When we provide sequelize with an include which has empty `.attributes` in it, it stops
            // there and doesn't go further in the include tree.
            // To ensure that no `.attributes` are not empty, we add the pk of each collection in the
            // projection.
            //
            // include: [{
            //   association: 'users',
            //   attributes: [],
            //   include: { association: 'posts', attributes: ['i_want_this_field'] }
            // }]
            projection.withPks(this), new datasource_toolkit_1.Projection().union(filter.conditionTree?.projection, filter.sort?.projection)),
            limit: filter.page?.limit,
            offset: filter.page?.skip,
            order: this.queryConverter.getOrderFromSort(filter.sort),
            subQuery: false,
        };
        const records = await this.model.findAll(query);
        const rawRecords = records.map(record => serializer_1.default.serialize(record.get({ plain: true })));
        // Use projection to filter out the unwanted primary keys that were added to the projection
        // so that sequelize can do its job.
        return projection.apply(rawRecords);
    }
    async update(caller, filter, patch) {
        const options = {
            where: await this.queryConverter.getWhereFromConditionTreeToByPassInclude(filter.conditionTree),
            fields: Object.keys(patch),
        };
        await (0, error_handler_1.default)('update', () => this.model.update(patch, options));
    }
    async delete(caller, filter) {
        const options = {
            where: await this.queryConverter.getWhereFromConditionTreeToByPassInclude(filter.conditionTree),
        };
        await (0, error_handler_1.default)('delete', () => this.model.destroy(options));
    }
    async aggregate(caller, filter, aggregation, limit) {
        let aggregationField = aggregation.field;
        let aggregationFieldSchema;
        if (aggregation.operation === 'Count' || !aggregationField) {
            aggregationField = '*';
        }
        else {
            aggregationField = this.aggregationUtils.quoteField(aggregationField);
            aggregationFieldSchema = datasource_toolkit_1.CollectionUtils.getFieldSchema(this, aggregation.field);
        }
        const aggregationFunction = this.fn(aggregation.operation.toUpperCase(), this.col(aggregationField));
        const aggregationAttribute = [
            aggregationFunction,
            this.aggregationUtils.aggregateFieldName,
        ];
        const { groups, attributes } = this.aggregationUtils.getGroupAndAttributesFromAggregation(aggregation.groups);
        const query = {
            attributes: [...attributes, aggregationAttribute],
            group: groups,
            where: this.queryConverter.getWhereFromConditionTree(filter.conditionTree),
            include: this.queryConverter.getIncludeFromProjection(new datasource_toolkit_1.Projection(), aggregation.projection.union(filter.conditionTree?.projection)),
            limit,
            order: [this.aggregationUtils.getOrder(aggregationFunction)],
            subQuery: false,
            raw: true,
        };
        const rows = await this.model.findAll(query);
        return this.aggregationUtils.computeResult(rows, aggregation.groups, !aggregationFieldSchema || aggregationFieldSchema?.columnType === 'Number');
    }
}
exports.default = SequelizeCollection;
