// @forestadmin/datasource-sql/dist/introspection/helpers/default-value-parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const sequelize_1 = require("sequelize");
class DefaultValueParser {
    constructor(dialect) {
        this.dialect = dialect;
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    parse(expression, columnType) {
        if (expression === null || expression === undefined)
            return undefined;
        if (typeof expression === 'string' && expression.startsWith('NULL'))
            return null;
        // FA backend not handle correctly
        if (columnType.type === 'array')
            return undefined;
        try {
            const result = this.parseGeneric(expression, columnType);
            return result !== undefined ? result : (0, sequelize_1.literal)(expression);
        }
        catch (e) {
            return (0, sequelize_1.literal)(expression);
        }
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    isLiteral(expression, columnType) {
        return this.parse(expression, columnType)?.constructor.name === 'Literal';
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    parseGeneric(expression, columnType) {
        let sanitizedExpression = expression;
        // sanitize string
        if (typeof expression === 'string') {
            sanitizedExpression = this.sanitizeExpression(expression);
        }
        if (columnType.type === 'enum') {
            return sanitizedExpression;
        }
        switch (columnType.subType) {
            case 'BOOLEAN':
                return [true, 'true', 'TRUE', "b'1'", '1'].includes(sanitizedExpression);
            case 'NUMBER':
            case 'BIGINT':
            case 'FLOAT':
            case 'DOUBLE':
                return this.parseNumber(sanitizedExpression);
            case 'DATE':
            case 'DATEONLY':
                return this.literalUnlessMatch(/^(\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2})|(\d{4}-\d{2}-\d{2})|(\d{2}:\d{2}:\d{2})$/, sanitizedExpression);
            case 'STRING':
                return sanitizedExpression;
            case 'JSON':
            case 'JSONB':
                return JSON.parse(sanitizedExpression);
            default:
                return undefined;
        }
    }
    sanitizeExpression(expression) {
        let sanitizedExpression = expression;
        if (/^'.*'$/.test(sanitizedExpression)) {
            sanitizedExpression = expression.substring(1, expression.length - 1);
        }
        if (this.dialect === 'mssql') {
            // Sequelize send default values with weird char at the beginning (`(Ndefault value`)
            sanitizedExpression = sanitizedExpression.replace(/\(N/, '');
            while (/^\(.*\)$/.test(sanitizedExpression)) {
                sanitizedExpression = sanitizedExpression.substring(1, sanitizedExpression.length - 1);
            }
        }
        return sanitizedExpression;
    }
    literalUnlessMatch(regexp, expression) {
        return regexp.test(expression) ? expression : (0, sequelize_1.literal)(expression);
    }
    parseNumber(expression) {
        const result = Number.parseFloat(expression);
        return Number.isNaN(result) ? (0, sequelize_1.literal)(expression) : result;
    }
}
exports.default = DefaultValueParser;
