// @forestadmin/datasource-sql/dist/connection/services/reverse-proxy.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const net_1 = __importDefault(require("net"));
const service_1 = __importDefault(require("./service"));
/**
 * ReverseProxy is used to redirect all the database requests.
 * Sequelize does not take a socket as an argument,
 * so we need to redirect all the traffic to a new socket.
 * This is done by creating a server that will listen on a random port.
 * We change the host and port of the Sequelize options to point to the reverse proxy.
 * The reverse proxy will then redirect all the traffic to the database through the tunnel.
 */
class ReverseProxy extends service_1.default {
    get host() {
        return this.server.address().address;
    }
    get port() {
        return this.server.address().port;
    }
    constructor() {
        super(null, null, null, null);
        this.server = net_1.default.createServer(this.connect.bind(this));
    }
    start() {
        return new Promise((resolve, reject) => {
            this.server.on('error', reject);
            // By using port 0, the operating system
            // will assign an available port for the server to listen on.
            this.server.listen(0, '127.0.0.1', resolve);
        });
    }
    async stop() {
        try {
            // close all the connected clients before closing the server
            await super.stop();
        }
        finally {
            await new Promise((resolve, reject) => {
                this.server.close(e => {
                    if (e)
                        reject(e);
                    else
                        resolve();
                });
            });
        }
    }
    async connect(socket) {
        try {
            this.addConnectedClient(socket);
            socket.on('close', () => this.destroySocketIfUnclosedAndSaveError(socket));
            socket.on('error', error => this.destroySocketIfUnclosedAndSaveError(socket, error));
            const tunnel = await super.connect();
            if (tunnel)
                tunnel.pipe(socket).pipe(tunnel);
            return tunnel;
        }
        catch (error) {
            this.destroySocketIfUnclosedAndSaveError(socket, error);
            // don't throw the error to avoid crashing the server because the error is already handled
        }
    }
}
exports.default = ReverseProxy;
