// @forestadmin/datasource-sql/dist/connection/services/service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Service {
    get error() {
        return this.errors.length > 0 ? this.errors[0] : null;
    }
    get debugUri() {
        return `${this.sourceHost}:${this.sourcePort}`;
    }
    get debugForwardUri() {
        return `${this.targetHost}:${this.targetPort}`;
    }
    constructor(sourceHost, sourcePort, targetHost, targetPort) {
        this.connectedClients = new Set();
        this.errors = [];
        this.sourceHost = sourceHost;
        this.sourcePort = sourcePort;
        this.targetHost = targetHost;
        this.targetPort = targetPort;
    }
    /** link a service */
    link(service) {
        this.linkedService = service;
    }
    /** stop the service by stopping the linked service and destroying all its clients. */
    async stop() {
        try {
            await this.linkedService?.stop();
        }
        finally {
            this.connectedClients.forEach(client => client.destroy());
        }
    }
    /** call the linked service connection callback */
    async connect(socket) {
        return (await this.linkedService?.connect(socket)) || socket;
    }
    /** destroy the given socket if it is not closed and save the error */
    destroySocketIfUnclosedAndSaveError(socket, error) {
        if (error)
            this.errors.push(error);
        if (socket) {
            if (!socket.closed)
                socket.destroy();
            this.connectedClients.delete(socket);
        }
    }
    /** register a socket as a client to be destroyed when the service is stopped */
    addConnectedClient(client) {
        this.connectedClients.add(client);
    }
}
exports.default = Service;
