// @forestadmin/datasource-sql/dist/connection/services/socks-proxy.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const socks_1 = require("socks");
const service_1 = __importDefault(require("./service"));
const errors_1 = require("../errors");
class SocksProxy extends service_1.default {
    constructor(proxyOptions, targetHost, targetPort) {
        super(proxyOptions.host, proxyOptions.port, targetHost, targetPort);
        this.options = proxyOptions;
    }
    async connect() {
        let socks5Client;
        try {
            socks5Client = await socks_1.SocksClient.createConnection({
                proxy: { ...this.options, type: 5 },
                command: 'connect',
                destination: { host: this.targetHost, port: this.targetPort },
                timeout: 4000,
            });
            this.addConnectedClient(socks5Client.socket);
            socks5Client.socket.on('close', () => this.destroySocketIfUnclosedAndSaveError(socks5Client.socket));
            socks5Client.socket.on('error', error => this.destroySocketIfUnclosedAndSaveError(socks5Client.socket, new errors_1.ProxyConnectError(error.message, this.debugUri)));
            const tunnel = await super.connect(socks5Client.socket);
            if (tunnel) {
                // destroy the proxy socket when the tunnel is closed or an error occurs
                // this is very important to avoid unclose database connections
                tunnel.on('close', () => this.destroySocketIfUnclosedAndSaveError(socks5Client.socket));
                tunnel.on('error', error => this.destroySocketIfUnclosedAndSaveError(socks5Client.socket, new errors_1.ProxyConnectError(error.message, this.debugUri)));
            }
            return tunnel;
        }
        catch (error) {
            let serviceError = new errors_1.ProxyConnectError(error.message, this.debugUri);
            if (error.message.includes('Socket closed') ||
                error.message.includes('Socks5 proxy rejected connection')) {
                serviceError = new errors_1.ProxyForwardError(error.message, this.debugForwardUri);
            }
            this.destroySocketIfUnclosedAndSaveError(socks5Client?.socket, serviceError);
            throw serviceError;
        }
    }
}
exports.default = SocksProxy;
