// @forestadmin/datasource-sql/dist/connection/services/ssh-tunnel.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ssh2_1 = require("ssh2");
const service_1 = __importDefault(require("./service"));
const errors_1 = require("../errors");
class SshTunnel extends service_1.default {
    constructor(options, targetHost, targetPort) {
        super(options.host, options.port, targetHost, targetPort);
        this.clients = new Set();
        this.options = options;
    }
    async stop() {
        try {
            await super.stop();
        }
        finally {
            this.clients.forEach(client => this.endClient(client));
        }
    }
    async connect(socket) {
        const client = new ssh2_1.Client();
        // list all the clients to be able to close them all when the service is stopped
        this.clients.add(client);
        try {
            return await new Promise((resolve, reject) => {
                client.on('error', e => reject(new errors_1.SshConnectError(e.message, this.debugUri)));
                client.on('ready', async () => {
                    try {
                        resolve(await this.buildTunnel(client));
                    }
                    catch (error) {
                        reject(error);
                    }
                });
                // connect to the SSH server
                // will trigger the 'ready' event if the connection is successful
                // if the connection fails, the 'error' event will be triggered
                client.connect({ ...this.options, sock: socket });
            });
        }
        catch (error) {
            this.endClient(client);
            this.errors.push(error);
            throw error;
        }
    }
    async buildTunnel(client) {
        // tell to the SSH server to forward all the traffic to the target host and port
        return new Promise((resolve, reject) => {
            // source host and port are not used by the SSH server
            client.forwardOut('', 0, this.targetHost, this.targetPort, async (error, stream) => {
                if (error)
                    return reject(new errors_1.SshForwardError(error.message, this.debugForwardUri));
                this.addConnectedClient(stream);
                stream.on('error', e => this.destroySocketIfUnclosedAndSaveError(stream, new errors_1.SshConnectError(e.message, this.debugUri)));
                stream.on('close', () => {
                    this.destroySocketIfUnclosedAndSaveError(stream);
                    this.endClient(client);
                });
                return resolve(await super.connect(stream));
            });
        });
    }
    endClient(client) {
        client.end();
        this.clients.delete(client);
    }
}
exports.default = SshTunnel;
