// @forestadmin/datasource-sql/dist/connection/errors.js
"use strict";
/* eslint-disable max-classes-per-file */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SshForwardError = exports.SshConnectError = exports.ProxyForwardError = exports.ProxyConnectError = exports.DatabaseConnectError = void 0;
class BaseError extends Error {
    constructor(message, debugUri, details) {
        super(details ? `${message}\n${details}` : message);
        this.name = this.constructor.name;
        this.details = details;
        this.debugUri = debugUri;
    }
}
class DatabaseConnectError extends BaseError {
    constructor(message, debugDatabaseUri, source = 'Database') {
        // remove tcp protocol because its not added by the user
        const sanitizedUri = debugDatabaseUri.replace('tcp://', '');
        super(`Unable to connect to the given uri: ${sanitizedUri}.`, sanitizedUri, message);
        this.source = source;
    }
}
exports.DatabaseConnectError = DatabaseConnectError;
class ProxyConnectError extends BaseError {
    constructor(message, debugProxyUri) {
        super(`Your proxy has encountered an error. Unable to connect to the given uri: ${debugProxyUri}.`, debugProxyUri, message);
        this.source = 'Proxy';
    }
}
exports.ProxyConnectError = ProxyConnectError;
class ProxyForwardError extends BaseError {
    constructor(message, debugForwardedUri) {
        super(`Your proxy forwarded connection has encountered an error.` +
            `Unable to connect to the given uri: ${debugForwardedUri}.`, debugForwardedUri, message);
        this.source = 'Proxy';
    }
}
exports.ProxyForwardError = ProxyForwardError;
class SshConnectError extends BaseError {
    constructor(message, debugSshUri, source = 'Ssh') {
        super(`Your ssh connection has encountered an error. ` +
            `Unable to connect to the given ssh uri: ${debugSshUri}`, debugSshUri, message);
        this.source = source;
    }
}
exports.SshConnectError = SshConnectError;
class SshForwardError extends BaseError {
    constructor(message, debugForwardedUri) {
        super(`Your ssh forward connection has encountered an error. ` +
            `Unable to connect to the given ssh uri: ${debugForwardedUri}`, debugForwardedUri, message);
        this.source = 'Ssh';
    }
}
exports.SshForwardError = SshForwardError;
