// @forestadmin/datasource-sql/dist/orm-builder/relations.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const relation_extractor_1 = __importDefault(require("./helpers/relation-extractor"));
const relation_name_generator_1 = __importDefault(require("./helpers/relation-name-generator"));
class RelationBuilder {
    static defineRelations(sequelize, logger, tables) {
        for (const table of tables) {
            this.defineTableRelations(sequelize, logger, table, tables);
        }
    }
    static defineTableRelations(sequelize, logger, table, tables) {
        const relations = relation_extractor_1.default.listRelations(table.name, tables);
        const relationNames = relation_name_generator_1.default.getUniqueRelationNames(table, relations);
        for (const [index, relation] of relations.entries()) {
            const as = relationNames[index];
            try {
                this.defineRelation(sequelize, relation, as);
            }
            catch (e) {
                logger?.('Warn', `Skipping relation "${table.name}.${as}" because of error: ${e.message}`);
            }
        }
    }
    static defineRelation(sequelize, relation, as) {
        const sourceModel = sequelize.model(relation.from);
        const targetModel = sequelize.model(relation.to);
        if (relation.type === 'BelongsTo') {
            sourceModel.belongsTo(targetModel, {
                as,
                foreignKey: relation.foreignKey,
                targetKey: relation.foreignKeyTarget,
            });
        }
        else if (relation.type === 'HasMany') {
            sourceModel.hasMany(targetModel, {
                as,
                foreignKey: relation.originKey,
                sourceKey: relation.originKeyTarget,
            });
        }
        else if (relation.type === 'BelongsToMany') {
            sourceModel.belongsToMany(targetModel, {
                as,
                through: relation.through,
                otherKey: relation.foreignKey,
                foreignKey: relation.originKey,
                targetKey: relation.foreignKeyTarget,
                sourceKey: relation.originKeyTarget,
            });
        }
        else if (relation.type === 'HasOne') {
            sourceModel.hasOne(targetModel, {
                as,
                foreignKey: relation.originKey,
                sourceKey: relation.originKeyTarget,
            });
        }
    }
}
exports.default = RelationBuilder;
