// @forestadmin/forestadmin-client/dist/auth/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const openid_client_1 = require("openid-client");
const errors_1 = require("./errors");
const server_1 = __importDefault(require("../utils/server"));
class AuthService {
    constructor(options) {
        this.options = options;
    }
    /**
     * Initialize the authentication client upfront. This speeds up the first
     * authentication request.
     */
    async init() {
        try {
            await this.createClient();
        }
        catch (e) {
            // Sometimes the authentication client can't be initialized because of a
            // server or network error. We don't want the application to crash.
            this.options.logger('Warn', 
            // eslint-disable-next-line max-len
            `Error while registering the authentication client. Authentication might not work: ${e.message}`);
        }
    }
    async getUserInfo(renderingId, accessToken) {
        const url = `/liana/v2/renderings/${renderingId}/authorization`;
        const headers = { 'forest-token': accessToken };
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const response = await server_1.default.query(this.options, 'get', url, headers);
        return {
            id: Number(response.data.id),
            email: response.data.attributes.email,
            firstName: response.data.attributes.first_name,
            lastName: response.data.attributes.last_name,
            team: response.data.attributes.teams[0],
            role: response.data.attributes.role,
            permissionLevel: response.data.attributes.permission_level,
            renderingId,
            tags: response.data.attributes.tags?.reduce((memo, { key, value }) => ({ ...memo, [key]: value }), {}),
        };
    }
    async generateAuthorizationUrl({ scope, state, }) {
        await this.createClient();
        return this.client.authorizationUrl({
            scope,
            state,
        });
    }
    async generateTokens({ query, state, }) {
        await this.createClient();
        try {
            const tokens = await this.client.callback(undefined, query, { state });
            return {
                accessToken: tokens.access_token,
            };
        }
        catch (e) {
            this.handleError(e);
        }
    }
    async createClient() {
        if (this.client)
            return;
        // We can't use async 'Issuer.discover' because the oidc config is behind an auth-wall.
        const url = '/oidc/.well-known/openid-configuration';
        const config = await server_1.default.query(this.options, 'get', url);
        const issuer = new openid_client_1.Issuer(config);
        const registration = { token_endpoint_auth_method: 'none' };
        this.client = await issuer.Client.register(registration, {
            initialAccessToken: this.options.envSecret,
        });
    }
    handleError(e) {
        if (e instanceof openid_client_1.errors.OPError) {
            throw new errors_1.AuthenticationError(e);
        }
        throw e;
    }
}
exports.default = AuthService;
